pkg/querier/worker/util.go

```diff
@@ -0,0 +1,80 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package worker
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"go.uber.org/atomic"
+)
+
+// executionContext wraps the context.Context used to run the querier's worker loop and execute
+// queries. The purpose of the execution context is to gracefully shutdown queriers, waiting
+// until inflight queries are terminated before the querier process exits.
+//
+// How it's used:
+// - The querier worker's loop run in a dedicated context, called the "execution context".
+// - The execution context is canceled when the worker context gets cancelled (ie. querier is shutting down)
+//   and there's no inflight query execution. In case there's an inflight query, the execution context is canceled
+//   once the inflight query terminates and the response has been sent.
+type executionContext struct {
+	execCtx       context.Context
+	execCancel    context.CancelFunc
+	inflightQuery *atomic.Bool
+}
+
+func newExecutionContext(workerCtx context.Context, logger log.Logger) *executionContext {
```

Comment:
Nit: It's not clear that one needs to call `cancel()` on returned execution context (documentation missing). Returning `ctx, cancelFn, *atomic.Bool()` directly (without wrapping to the struct) would be more consistent with existing context-related functions, and more explicit that cancelFn must be called.

Reply:
I added a comment on `newExecutionContext()`. About removing the struct I will see how I will end up fixing the same for the case Mimir is running without query-scheduler first.